cmd/kubeadm/app/cmd/config.go

```diff
@@ -118,36 +116,80 @@ func NewCmdConfigPrintDefault(out io.Writer) *cobra.Command {
-			API:               kubeadmapiv1alpha3.API{AdvertiseAddress: "1.2.3.4"},
-			BootstrapTokens:   []kubeadmapiv1alpha3.BootstrapToken{sillyToken},
-			KubernetesVersion: fmt.Sprintf("v1.%d.0", constants.MinimumControlPlaneVersion.Minor()+1),
-		})
+		return getDefaultInitConfigBytes()
+
 	case constants.NodeConfigurationKind:
-		internalcfg, err = configutil.NodeConfigFileAndDefaultsToInternalConfig("", &kubeadmapiv1alpha3.NodeConfiguration{
-			Token: sillyToken.Token.String(),
-			DiscoveryTokenAPIServers:               []string{"kube-apiserver:6443"},
-			DiscoveryTokenUnsafeSkipCAVerification: true,
-		})
-		// TODO: DiscoveryTokenUnsafeSkipCAVerification: true needs to be set for validation to pass, but shouldn't be recommended as the default
+		return getDefaultNodeConfigBytes()
+
 	default:
-		err = fmt.Errorf("--api-object needs to be one of %v", availableAPIObjects)
+		// Is this a component config?
+		registration, ok := componentconfigs.Known[componentconfigs.RegistrationKind(apiObject)]
+		if !ok {
+			return []byte{}, fmt.Errorf("--api-object needs to be one of %v", getAvailableAPIObjects())
+		}
+		return getDefaultComponentConfigBytes(registration)
+	}
+}
+
+func getAvailableAPIObjects() []string {
+	objects := []string{constants.InitConfigurationKind, constants.MasterConfigurationKind, constants.NodeConfigurationKind}
+	for componentType := range componentconfigs.Known {
+		objects = append(objects, string(componentType))
 	}
+	return objects
+}
+
+func getDefaultedInitConfig() (*kubeadmapi.InitConfiguration, error) {
+	return configutil.ConfigFileAndDefaultsToInternalConfig("", &kubeadmapiv1alpha3.InitConfiguration{
+		API:               kubeadmapiv1alpha3.API{AdvertiseAddress: "1.2.3.4"},
+		BootstrapTokens:   []kubeadmapiv1alpha3.BootstrapToken{sillyToken},
+		KubernetesVersion: fmt.Sprintf("v1.%d.0", constants.MinimumControlPlaneVersion.Minor()+1),
+	})
+}
+
+func getDefaultInitConfigBytes() ([]byte, error) {
+	internalcfg, err := getDefaultedInitConfig()
 	if err != nil {
 		return []byte{}, err
 	}
+
 	return configutil.MarshalKubeadmConfigObject(internalcfg)
 }
 
+func getDefaultNodeConfigBytes() ([]byte, error) {
+	internalcfg, err := configutil.NodeConfigFileAndDefaultsToInternalConfig("", &kubeadmapiv1alpha3.NodeConfiguration{
+		Token: sillyToken.Token.String(),
+		DiscoveryTokenAPIServers:               []string{"kube-apiserver:6443"},
+		DiscoveryTokenUnsafeSkipCAVerification: true, // TODO: DiscoveryTokenUnsafeSkipCAVerification: true needs to be set for validation to pass, but shouldn't be recommended as the default
+	})
+	if err != nil {
+		return []byte{}, err
+	}
+
+	return configutil.MarshalKubeadmConfigObject(internalcfg)
+}
+
+func getDefaultComponentConfigBytes(registration componentconfigs.Registration) ([]byte, error) {
+	defaultedInitConfig, err := getDefaultedInitConfig()
+	if err != nil {
+		return []byte{}, err
+	}
+
+	realobj, ok := registration.GetFromInternalConfig(defaultedInitConfig)
```

Comment:
i haven't look at how this works or tested it.
so we use the init configuration from `getDefaultedInitConfig()` and then pass that to `registration.GetFromInternalConfig()` to be able to get the config for a certain known component type?

it seems a bit strange as an interface. :thinking:

Reply:
This is how it works. Component configs can be directly influenced by a top level config (currently `InitConfiguration`). For example an option in some component config may be directly depending on the value of another option in the `InitConfiguration`